File: taurus-protect-sdk-typescript/src/models/price.ts
/**
 * Price models for Taurus-PROTECT SDK.
 *
 * Provides domain models for cryptocurrency prices, price history,
 * and currency conversion operations.
 */

import type { Currency } from './currency';
import type { CursorPage, CursorPageOptions } from './pagination';

/**
 * Represents a currency price/exchange rate.
 *
 * Prices track exchange rates between currencies (e.g., ETH to USD) along with
 * metadata about the rate source, precision, and price changes. Prices are
 * used for portfolio valuation, transaction display, and reporting.
 */
export interface PriceSignature {
  /** The signing user's ID */
  readonly userId?: string;
  /** Base64-encoded ECDSA signature over the canonical price form */
  readonly signature?: string;
}

export interface Price {
  /** The blockchain this price applies to (e.g., "ethereum", "bitcoin") */
  readonly blockchain?: string;
  /** The source currency code being converted from (e.g., "ETH", "BTC") */
  readonly currencyFrom?: string;
  /** The target currency code being converted to (e.g., "USD", "EUR") */
  readonly currencyTo?: string;
  /** The number of decimal places for the rate precision */
  readonly decimals?: string;
  /** The exchange rate value (amount of currencyTo per unit of currencyFrom) */
  readonly rate?: string;
  /** PRICEUPDATER signatures over the canonical price form */
  readonly signatures?: PriceSignature[];
  /** The percentage price change over the last 24 hours */
  readonly changePercent24Hour?: string;
  /** The data source for this price (e.g., "coingecko", "cryptocompare") */
  readonly source?: string;
  /** Timestamp when this price record was created */
  readonly createdAt?: Date;
  /** Timestamp when this price record was last updated */
  readonly updatedAt?: Date;
  /** Detailed information about the source currency */
  readonly currencyFromInfo?: Currency;
  /** Detailed information about the target currency */
  readonly currencyToInfo?: Currency;
}

/**
 * Represents a single data point in a price history timeline (OHLCV candlestick data).
 *
 * Price history points capture OHLCV (Open, High, Low, Close, Volume) data for a specific
 * time period, enabling historical price analysis and charting. This follows standard
 * financial candlestick data format used in trading and portfolio analysis.
 */
export interface PriceHistoryPoint {
  /** The start timestamp of the time period this data point covers */
  readonly periodStartDate?: Date;
  /** The blockchain this price data applies to (e.g., "ethereum", "bitcoin") */
  readonly blockchain?: string;
  /** The source currency code being converted from (e.g., "ETH", "BTC") */
  readonly currencyFrom?: string;
  /** The target currency code being converted to (e.g., "USD", "EUR") */
  readonly currencyTo?: string;
  /** The highest price during this time period */
  readonly high?: string;
  /** The lowest price during this time period */
  readonly low?: string;
  /** The opening price at the start of this time period */
  readonly open?: string;
  /** The closing price at the end of this time period */
  readonly close?: string;
  /** The trading volume in the source currency during this period */
  readonly volumeFrom?: string;
  /** The trading volume in the target currency during this period */
  readonly volumeTo?: string;
  /** The percentage price change during this time period */
  readonly changePercent?: string;
  /** Detailed information about the source currency */
  readonly currencyFromInfo?: Currency;
  /** Detailed information about the target currency */
  readonly currencyToInfo?: Currency;
}

/**
 * Represents the result of a currency conversion operation.
 *
 * Conversion results contain both the raw value and a human-readable main unit value
 * (accounting for decimals), along with currency metadata. This is typically used
 * when displaying converted amounts in the UI or for reporting purposes.
 */
export interface ConversionResult {
  /** The currency symbol for display (e.g., "USD", "EUR", "ETH") */
  readonly symbol?: string;
  /** The converted value in the smallest unit (e.g., wei for ETH, satoshi for BTC) */
  readonly value?: string;
  /** The converted value in main units (e.g., ETH instead of wei), formatted for display */
  readonly mainUnitValue?: string;
  /** Detailed information about the target currency */
  readonly currencyInfo?: Currency;
}

/**
 * Options for getting price history.
 */
export interface GetPriceHistoryOptions {
  /** The base currency (e.g., "ETH", "BTC") */
  base: string;
  /** The quote currency (e.g., "USD", "EUR") */
  quote: string;
  /** Number of daily points to return, newest first: 1-365 (default 20) */
  limit?: number;
}

/**
 * Options for listing prices. A cursor list: `pageSize` 1-100 (default 20) and the
 * `cursor` of a previous page.
 *
 * `fromCurrencyId` alone keeps the prices from that currency, `toCurrencyIds` alone the
 * prices into those currencies, and both together the prices from the one into the
 * others.
 */
export interface ListPricesOptions extends CursorPageOptions {
  /** Keep only the primary price of each currency pair */
  readonly onlyPrimary?: boolean;
  /** Sort order ("ASC" or "DESC") */
  readonly sortOrder?: string;
  /** Keep prices whose source currency is this currency ID */
  readonly fromCurrencyId?: string;
  /** Keep prices whose target currency is one of these currency IDs */
  readonly toCurrencyIds?: string[];
}

/**
 * A page of verified prices.
 */
export interface ListPricesResult {
  /** The prices of this page, each verified against its signature */
  readonly items: Price[];
  /** Cursor pagination: continue with `cursor: pagination.nextCursor` while `hasMore` */
  readonly pagination: CursorPage;
}

/**
 * Options for currency conversion.
 */
export interface ConvertOptions {
  /** The source currency (e.g., "ETH", "BTC") */
  currency: string;
  /** The amount to convert (as a string to preserve precision) */
  amount: string;
  /** Optional list of currency symbols to convert to */
  symbols?: string[];
  /** Optional list of target currency IDs to convert to */
  targetCurrencyIds?: string[];
}

File: taurus-protect-sdk-go/pkg/protect/model/price.go
package model

import "time"

// Price represents a currency price with exchange rate information.
type Price struct {
	// Blockchain is the blockchain network.
	Blockchain string `json:"blockchain,omitempty"`
	// CurrencyFrom is the base currency symbol (e.g., "BTC").
	CurrencyFrom string `json:"currency_from,omitempty"`
	// CurrencyTo is the quote currency symbol (e.g., "USD").
	CurrencyTo string `json:"currency_to,omitempty"`
	// Decimals is the number of decimal places for the rate.
	Decimals string `json:"decimals,omitempty"`
	// Rate is the exchange rate between the currencies.
	Rate string `json:"rate,omitempty"`
	// Signatures contains price signatures for verification.
	Signatures []PriceSignature `json:"signatures,omitempty"`
	// ChangePercent24Hour is the 24-hour price change percentage.
	ChangePercent24Hour string `json:"change_percent_24hour,omitempty"`
	// Source is the price data source.
	Source string `json:"source,omitempty"`
	// CreationDate is when the price was created.
	CreationDate time.Time `json:"creation_date,omitempty"`
	// UpdateDate is when the price was last updated.
	UpdateDate time.Time `json:"update_date,omitempty"`
	// CurrencyFromInfo contains detailed information about the base currency.
	CurrencyFromInfo *CurrencyInfo `json:"currency_from_info,omitempty"`
	// CurrencyToInfo contains detailed information about the quote currency.
	CurrencyToInfo *CurrencyInfo `json:"currency_to_info,omitempty"`
	// ID is the price's identifier.
	ID string `json:"id,omitempty"`
	// IsPrimary marks the primary price of its currency pair.
	IsPrimary bool `json:"is_primary,omitempty"`
	// Status is the deviation status of a primary price.
	Status string `json:"status,omitempty"`
}

// PriceSignature represents a signature for price verification.
type PriceSignature struct {
	// UserID is the ID of the user who signed the price.
	UserID string `json:"user_id,omitempty"`
	// Signature is the cryptographic signature.
	Signature string `json:"signature,omitempty"`
}

// PriceHistoryPoint represents a single point in price history (OHLCV data).
type PriceHistoryPoint struct {
	// PeriodStartDate is the start time of the period.
	PeriodStartDate time.Time `json:"period_start_date,omitempty"`
	// Blockchain is the blockchain network.
	Blockchain string `json:"blockchain,omitempty"`
	// CurrencyFrom is the base currency symbol.
	CurrencyFrom string `json:"currency_from,omitempty"`
	// CurrencyTo is the quote currency symbol.
	CurrencyTo string `json:"currency_to,omitempty"`
	// High is the highest price in the period.
	High string `json:"high,omitempty"`
	// Low is the lowest price in the period.
	Low string `json:"low,omitempty"`
	// Open is the opening price of the period.
	Open string `json:"open,omitempty"`
	// Close is the closing price of the period.
	Close string `json:"close,omitempty"`
	// VolumeFrom is the trading volume in the base currency.
	VolumeFrom string `json:"volume_from,omitempty"`
	// VolumeTo is the trading volume in the quote currency.
	VolumeTo string `json:"volume_to,omitempty"`
	// ChangePercent is the price change percentage for the period.
	ChangePercent string `json:"change_percent,omitempty"`
	// CurrencyFromInfo contains detailed information about the base currency.
	CurrencyFromInfo *CurrencyInfo `json:"currency_from_info,omitempty"`
	// CurrencyToInfo contains detailed information about the quote currency.
	CurrencyToInfo *CurrencyInfo `json:"currency_to_info,omitempty"`
}

// ConversionValue represents a converted amount in a target currency.
type ConversionValue struct {
	// Symbol is the currency symbol.
	Symbol string `json:"symbol,omitempty"`
	// Value is the value in the smallest currency unit.
	Value string `json:"value,omitempty"`
	// MainUnitValue is the value in the main currency unit.
	MainUnitValue string `json:"main_unit_value,omitempty"`
	// CurrencyInfo contains detailed information about the currency.
	CurrencyInfo *CurrencyInfo `json:"currency_info,omitempty"`
}

// ConversionResult represents the result of a currency conversion.
type ConversionResult struct {
	// CurrencyFrom is the source currency symbol.
	CurrencyFrom string `json:"currency_from,omitempty"`
	// BaseCurrency is the base currency symbol.
	BaseCurrency string `json:"base_currency,omitempty"`
	// Values contains the converted amounts for each target currency.
	Values []*ConversionValue `json:"values,omitempty"`
	// FullCurrencyFrom contains detailed information about the source currency.
	FullCurrencyFrom *CurrencyInfo `json:"full_currency_from,omitempty"`
	// FullBaseCurrency contains detailed information about the base currency.
	FullBaseCurrency *CurrencyInfo `json:"full_base_currency,omitempty"`
}

// ListPricesOptions filters and pages PriceService.ListPrices.
//
// The currency filter is one of three shapes, chosen by which fields are set: FromCurrencyID
// alone (prices from that currency), FromCurrencyID with ToCurrencyIDs (from it to those), or
// ToCurrencyIDs alone (prices to those currencies). Neither sets no currency filter.
type ListPricesOptions struct {
	// OnlyPrimary keeps only the primary price of each currency pair.
	OnlyPrimary bool
	// SortOrder is ASC or DESC (the server's default).
	SortOrder string
	// FromCurrencyID is the currency the prices convert from.
	FromCurrencyID string
	// ToCurrencyIDs are the currencies the prices convert to.
	ToCurrencyIDs []string
	// PageSize is the page size: 0 selects DefaultPageSize, above MaxPageSize is an error.
	PageSize int64
	// Cursor is a previous page's Page.NextCursor; the SDK then requests the NEXT page.
	Cursor string
}

// ListPricesResult is one page of verified prices.
type ListPricesResult struct {
	// BaseCurrency is the tenant's base currency.
	BaseCurrency string `json:"base_currency,omitempty"`
	// Prices is the list of currency prices, each signature-verified.
	Prices []*Price `json:"prices"`
	// Page continues the list: pass Page.NextCursor as the next Cursor until HasMore is false.
	Page CursorPage `json:"page"`
}

// ConvertOptions contains options for the Convert method.
type ConvertOptions struct {
	// Currency is the source currency to convert from (required).
	Currency string
	// Amount is the amount to convert (required).
	Amount string
	// Symbols filters the target currencies by symbol.
	Symbols []string
	// TargetCurrencyIds filters the target currencies by ID.
	TargetCurrencyIds []string
}

// GetPriceHistoryOptions contains options for the GetPriceHistory method.
type GetPriceHistoryOptions struct {
	// Base is the base currency symbol (required).
	Base string
	// Quote is the quote currency symbol (required).
	Quote string
	// Limit is the number of newest daily points to return: 0 selects DefaultPageSize, above
	// MaxPriceHistoryLimit is an error. Price history cannot page.
	Limit int64
}

// GetPriceHistoryResult contains the result of a GetPriceHistory call.
type GetPriceHistoryResult struct {
	// History is the list of price history points.
	History []*PriceHistoryPoint `json:"history"`
	// Period is the time period between history points.
	Period string `json:"period,omitempty"`
}

// ExportPriceHistoryOptions contains options for the ExportPriceHistory method.
type ExportPriceHistoryOptions struct {
	// CurrencyPairs is the list of currency pairs to export (e.g., ["BTC/USD", "ETH/USD"]).
	CurrencyPairs []string
	// Limit is the number of newest points to export: 0 selects DefaultPageSize; there is no SDK
	// maximum (the server bounds it). The export cannot page.
	Limit int64
	// Format is the export format ("csv" or "json").
	Format string
}

// ExportPriceHistoryResult contains the result of an ExportPriceHistory call.
type ExportPriceHistoryResult struct {
	// Data contains the exported data in the requested format.
	Data string `json:"data,omitempty"`
	// Period is the time period between history points.
	Period string `json:"period,omitempty"`
}
